main/twai.c

```diff
@@ -18,24 +18,27 @@ void twai_receive_task(void *arg)
         twai_message_t twai_rx_msg;
         twai_receive(&twai_rx_msg, portMAX_DELAY); // If no message available, should block and yield.
         ESP_LOGI(TWAI_TAG, "Received TWAI message with identifier %08X and length %08X", twai_rx_msg.identifier, twai_rx_msg.data_length_code);
-        for (int i = 0; i < twai_rx_msg.data_length_code; i++) {
-            ESP_LOGD(TWAI_TAG, "RX Data: %02X", twai_rx_msg.data[i]);
-        }
-        for (int i = 0; i < NUM_ISOTP_LINK_CONTAINERS; ++i) {
-            IsoTpLinkContainer *isotp_link_container = &isotp_link_containers[i];
-            // flipped?
-            if (twai_rx_msg.identifier == isotp_link_container->link.receive_arbitration_id) {
-                ESP_LOGD(TWAI_TAG, "twai_receive_task: link match");
-                ESP_LOGD(TWAI_TAG, "Taking isotp_mutex");
-                xSemaphoreTake(isotp_mutex, (TickType_t)100);
-                ESP_LOGD(TWAI_TAG, "Took isotp_mutex");
-                isotp_on_can_message(&isotp_link_container->link, twai_rx_msg.data, twai_rx_msg.data_length_code);
-                ESP_LOGD(TWAI_TAG, "twai_receive_task: giving isotp_mutex");
-                xSemaphoreGive(isotp_mutex);
-                ESP_LOGD(TWAI_TAG, "twai_receive_task: giving wait_for_isotp_data_sem");
-                xSemaphoreGive(isotp_link_container->wait_for_isotp_data_sem);
+        if (esp_log_level_get('*') == ESP_LOG_DEBUG) {
```

Comment:
really hot path when connected to a car, seemed worthwhile to optimize at this level. might be silly. does compiler remove ESP_LOGD calls entirely if not configured to log debug level logs?

Reply:
The LOGD macro does introduce a runtime comparison (but not a function call), you can see how it's defined here. More work would be required to see if it optimizes out.

https://github.com/espressif/esp-idf/blob/8131d6f46d690e9cb60cc5cd457863cc5479351f/components/log/include/esp_log.h#L345

and here:

https://github.com/espressif/esp-idf/blob/8131d6f46d690e9cb60cc5cd457863cc5479351f/components/log/include/esp_log.h#L439 

We'd have to look at the generated object code to learn more. I'm probably... 80% sure that the ESP_LOGD way is going to be removed completely during an optimizer pass during compilation, since it compares a constant to a constant. With that said, this depends on *a lot* of factors so observing the behavior (looking at the dissassembly of a simple project with just a LOGD in it) would probably be the fastest/easiest way to figure it out.